utils.py

```diff
@@ -275,25 +273,93 @@ def set(self, lru_cache):
 ip_cache = LRUCacheSingleton()
 
 
-def is_ip_per_task(app):
-    """
-    Return whether the application is using IP-per-task.
-    :param app:  The application to check.
-    :return:  True if using IP per task, False otherwise.
-    """
-    return app.get('ipAddress') is not None
+def get_app_networking_mode(app):
+    mode = 'host'
 
+    if app.get('ipAddress'):
+        mode = 'container'
 
-def is_user_network(app):
-    """
-    Returns True if container network mode is set to USER
-    :param app:  The application to check.
-    :return:  True if using USER network, False otherwise.
-    """
-    c = app.get('container', {})
-    return c is not None and c.get('type', '') == 'DOCKER' and \
-        c.get('docker', {})\
-        .get('network', '') == 'USER'
+    _mode = app.get('container', {})\
+               .get('docker', {})\
+               .get('network', '')
+    if _mode == 'USER':
+        mode = 'container'
+    elif _mode == 'BRIDGE':
+        mode = 'container/bridge'
+
+    networks = app.get('networks', [])
+    for n in networks:
+        # Modes cannot be mixed, so assigning the last mode is fine
+        mode = n.get('mode', 'host')
+
+    return mode
+
+
+def get_task_ip(task, mode):
+    if mode == 'container':
+        task_ip_addresses = task.get('ipAddresses', [])
+        if not task_ip_addresses:
+            logger.warning("Task %s does not yet have an ip address allocated",
+                           task['id'])
+            return None
+        task_ip = task_ip_addresses[0].get('ipAddress')
+        if not task_ip:
+            logger.warning("Task %s does not yet have an ip address allocated",
+                           task['id'])
+            return None
+        return task_ip
+    else:
+        host = task.get('host')
+        if not host:
+            logger.warning("Could not find task host, ignoring")
+        task_ip = resolve_ip(host)
+        if not task_ip:
+            logger.warning("Could not resolve ip for host %s, ignoring",
+                           host)
+        return task_ip
+
+
+def get_app_port_mappings(app):
+    portMappings = app.get('container', {})\
+                      .get('docker', {})\
+                      .get('portMappings')
+    if portMappings:
+        return portMappings
+
+    portMappings = app.get('container', {})\
+                      .get('portMappings')
+    return portMappings
+
+
+def get_app_task_ports(app, task, mode):
+    if mode == 'host' or mode == 'container/bridge':
+        task_ports = task.get('ports')
+        if task_ports:
+            return task_ports
+        port_definitions = app.get('portDefinitions', [])
+        task_ports = [p['port']
+                      for p in port_definitions
+                      if 'port' in p]
+        if len(task_ports) == 0:
+            return None
```

Comment:
if `get_app_task_ports` is only called for marathon 1.4.x JSONs it might be nice to document that assumption